server/bin/update-server-xml.py

```diff
@@ -57,27 +57,70 @@ def _add_attributes(self, node, attributes):
             logger.debug("Setting %s to %s on %s" % (k, v, node.name))
             node.setProp(k, v)
 
+    def _is_different_impl(self, cnode, nnode):
+        logger.debug("Comparing nodes: %s and %s" % (cnode.name, nnode.name))
+
+        # Check if the nodes have different names...
+        if cnode.name != nnode.name:
+            logger.debug("Node names differ: %s != %s" % (cnode.name, nnode.name))
+            return True
+
+        # Compare attributes immediately present on both nodes...
+        cattribs = {}
+        nattribs = {}
+
+        if cnode.properties:
+            cattribs = { np.name: np.content for np in cnode.properties if np.type == "attribute" }
+
+        if nnode.properties:
+            nattribs = { np.name: np.content for np in nnode.properties if np.type == "attribute" }
+
+        if cattribs != nattribs:
+            logger.debug("Attributes differ: %s != %s" % (cattribs, nattribs))
+            return True
+
+        # Compare children...
+        # Impl note: depending on order is "safe" here, since a different order implies something changed
+        # either in what this config editor generates, or the end-user has modified the server.xml
+        cchildren = []
+        nchildren = []
+
+        if cnode.children:
+            for child in cnode.children:
+                cchildren.append(child)
+
+        if nnode.children:
+            for child in nnode.children:
+                nchildren.append(child)
+
+        if len(cchildren) != len(nchildren):
+            return True
+
+        for i in range(0, len(cchildren)):
+            if self._is_different_impl(cchildren[i], nchildren[i]):
+                logger.debug("Children differ: %s != %s" % (cchildren[i].name, nchildren[i].name))
+                return True
+
+        # The things we care about line up, no difference
+        logger.debug("Nodes are the same")
+        return False
+
     def _is_different(self, node):
-        current_attributes = {}
-        for property in node.properties:
-            if property.type == "attribute":
-                current_attributes[property.name] = property.content
-        return node.name != self.new_node.name or current_attributes != dict(self.attributes)
+        return self._is_different_impl(node, self.new_node)
 
     def _update(self, existing_nodes):
         different_nodes = filter(self._is_different, existing_nodes)
```

Comment:
Maybe instead of having a wrapper `_is_different` and `_is_different_impl`, we can just adjust this filter statement to

```
different_nodes = filter(lambda x: self._is_different(x, self.new_node), existing_nodes)
```

And then have the body of `_is_different` be replaced with the body of the current `_is_different_imple`.

Reply:
Doesn't that just offload the wrapping to the caller instead of the implementation?

Reply:
Yes, but I thought it was a little weird to have both `is_different` and `is_different_impl` methods.  Both times when reviewing, I got to `is_different` and sat scratching my head for a minute wondering "why is this even here?"

Reply:
Hrmm... Would you be okay if I moved the impl to an inner function within is_different (assuming calling conventions allow it)?

Reply:
Or, now that I think of it, a "static" or global function that handles the checking of two arbitrary nodes, and the method which calls into it to properly separate/name the two variations.

Reply:
> Or, now that I think of it, a "static" or global function that handles the checking of two arbitrary nodes, and the method which calls into it to properly separate/name the two variations.

Sure.

Reply:
Done.